creator-node/src/services/stateMachineManager/stateMonitoring/stateMonitoringUtils.js

```diff
@@ -391,10 +397,102 @@ const _aggregateOps = async (
+ * @param {number} param.secondaryClock clock value on user's secondary
+ * @param {string} param.primaryFilesHash filesHash on user's primary
+ * @param {string} param.secondaryFilesHash filesHash on user's secondary
+ * @returns {SYNC_MODES} syncMode one of None, SyncSecondaryFromPrimary, SyncPrimaryFromSecondary
+ */
+const computeSyncModeForUserAndReplica = async ({
+  wallet,
+  primaryClock,
+  secondaryClock,
+  primaryFilesHash,
+  secondaryFilesHash
+}) => {
+  if (
+    !Number.isInteger(primaryClock) ||
+    !Number.isInteger(secondaryClock) ||
+    // `null` is a valid filesHash value; `undefined` is not
+    primaryFilesHash === undefined ||
+    secondaryFilesHash === undefined
+  ) {
+    throw new Error(
+      '[computeSyncModeForUserAndReplica] Error: Missing or invalid params'
+    )
+  }
+
+  if (
+    primaryClock === secondaryClock &&
+    primaryFilesHash === secondaryFilesHash
+  ) {
+    /**
+     * Nodes have identical data -> no sync needed
+     */
+    return SYNC_MODES.None
+  } else if (
+    primaryClock === secondaryClock &&
+    primaryFilesHash !== secondaryFilesHash
+  ) {
+    /**
+     * If clocks are same but filesHashes are not, this means secondary and primary states for user
+     *    have diverged. To fix this issue, primary should sync content from secondary and
+     *    subsequently force secondary to resync state from primary.
+     */
+    return SYNC_MODES.SyncPrimaryFromSecondary
+  } else if (primaryClock < secondaryClock) {
+    /**
+     * Secondary has more data than primary -> primary must sync from secondary
+     */
+    return SYNC_MODES.SyncPrimaryFromSecondary
+  } else if (primaryClock > secondaryClock && secondaryFilesHash === null) {
+    /**
+     * secondaryFilesHash will be null if secondary has no files for user -> secondary must sync from primary
+     */
+    return SYNC_MODES.SyncSecondaryFromPrimary
+  } else if (primaryClock > secondaryClock && secondaryFilesHash !== null) {
+    /**
+     * If primaryClock > secondaryClock, need to check that nodes have same content for each clock value. To do this, we compute filesHash from primary matching clock range from secondary.
+     */
+
+    let primaryFilesHashForRange
+    try {
+      // Throws error if failure after all retries
+      primaryFilesHashForRange = await retry(
+        async () =>
+          DBManager.fetchFilesHashFromDB({
+            lookupKey: { lookupWallet: wallet },
+            clockMin: 0,
+            clockMax: secondaryClock + 1
+          }),
+        { retries: FETCH_FILES_HASH_NUM_RETRIES }
+      )
+    } catch (e) {
+      const errorMsg = `[computeSyncModeForUserAndReplica] Error: failed DBManager.fetchFilesHashFromDB() - ${e.message}`
+      logger.error(errorMsg)
+      throw new Error(errorMsg)
+    }
+
+    if (primaryFilesHashForRange === secondaryFilesHash) {
+      return SYNC_MODES.SyncSecondaryFromPrimary
+    } else {
+      return SYNC_MODES.SyncPrimaryFromSecondary
```

Comment:
also curious about how this would be reached and if it's supposed to be reached at all or if it indicates some kind of unexpected error state that we should be logging or adding a metric for

Reply:
yep, added comment lmk if that clears this up

and 100% you're right, this would be great to add logging and metrics
this condition is logged already in `findSyncRequests.jobProcessor` ([linnk here](https://github.com/AudiusProject/audius-protocol/blob/ss-files-hash-bugfix-v3/creator-node/src/services/stateMachineManager/stateMonitoring/findSyncRequests.jobProcessor.js#L253))

[made a note to add metrics in follow up PR](https://linear.app/audius/issue/CON-236/pass-syncmode-through-job-data-add-metrics-by-syncmode)

Reply:
perfect 👍